xml_struct_derive/src/serialize/codegen.rs

```diff
@@ -0,0 +1,588 @@
+
+                Ok(())
+            }
+        }
+
+        #attr_impl
+    }
+}
+
+/// The sets of tokens which make up the implementations or bodies of
+/// `XmlSerialize` and `XmlSerializeAttr` trait methods.
+pub(super) struct ImplTokenSets {
+    /// The implementation of `XmlSerialize::serialize_as_element()` if it is
+    /// necessary to override the provided default implementation.
+    as_element_impl: TokenStream,
+
+    /// The body of `XmlSerialize::serialize_child_nodes()`.
+    child_nodes_body: TokenStream,
+
+    /// The body of `XmlSerializeAttr::serialize_as_attribute()` if the type is
+    /// capable of being serialized as such.
+    as_attr_body: Option<TokenStream>,
+}
+
+/// Generates the sets of tokens necessary for serializing a struct with the
+/// provided fields.
+pub(super) fn with_struct_fields(
+    fields: Vec<Field>,
+) -> impl FnOnce(&[XmlAttribute]) -> ImplTokenSets {
+    move |namespace_attrs| {
+        let Fields {
+            attr_fields,
+            child_fields,
+        } = partition_fields(fields);
+
+        let content_call = if !child_fields.is_empty() {
+            Some(quote! {
+                <Self as XmlSerialize>::serialize_child_nodes(self, writer)?;
+            })
+        } else {
+            None
+        };
+
+        let impl_body =
+            generate_xml_tag_calls(quote!(name), namespace_attrs, &attr_fields, content_call);
+
+        ImplTokenSets {
+            as_element_impl: quote! {
+                fn serialize_as_element<W: std::io::Write>(
+                    &self,
+                    writer: &mut ::quick_xml::writer::Writer<W>,
+                    name: &str,
+                ) -> Result<(), ::xml_struct::Error> {
+                    #impl_body
+
+                    Ok(())
+                }
+            },
+            child_nodes_body: generate_child_node_calls_for_fields(child_fields),
+
+            // Only unit-only enums can be serialized as attribute values.
+            as_attr_body: None,
+        }
+    }
+}
+
+/// Generates the sets of tokens necessary for serializing a unit enum as text
+/// nodes or attribute values.
+pub(super) fn with_text_variants(
+    variants: Vec<Ident>,
+) -> impl FnOnce(&[XmlAttribute]) -> ImplTokenSets {
+    // The property parsing code should guarantee that we have no namespace
+    // attributes to apply, since they don't make sense for text.
+    move |_| {
+        let match_arms: Vec<_> = variants
+            .iter()
+            .map(|variant| quote!(Self::#variant => stringify!(#variant)))
+            .collect();
+
+        let text_from_value = quote! {
+            let text = match self {
+                #(#match_arms,)*
+            };
+        };
+
+        ImplTokenSets {
+            // No namespaces can be declared on enums which are serialized as
+            // text, nor can they contain any attribute fields, so the default
+            // "as element" implementation is sufficient.
```

Comment:
suggestion: refer to the actual trait method where someone perusing the code can find said default implementation

```suggestion
            // text, nor can they contain any attribute fields, so the default
            // implementation for `XmlSerialize::serialize_as_element` is sufficient.
```